examples/benchmarking_comparison_models.py

```diff
@@ -0,0 +1,209 @@
+import art
+import asyncio
+from pydantic import BaseModel
+import re
+from art.utils import iterate_dataset
+from openai.types.chat import ChatCompletion
+import os
+
+PROJECT_NAME = "benchmarking-comparison-models"
+
+
+# This is a project-specific definition of a scenario you might use for training
+class MyScenarioSpec(BaseModel):
+    input_1: int
+    input_2: int
+    output: int
+
+
+training_scenarios = [
+    MyScenarioSpec(input_1=1, input_2=2, output=3),
+    MyScenarioSpec(input_1=2, input_2=3, output=5),
+    MyScenarioSpec(input_1=3, input_2=4, output=7),
+]
+
+test_scenarios = [
+    MyScenarioSpec(input_1=4, input_2=5, output=9),
+    MyScenarioSpec(input_1=5, input_2=6, output=11),
+]
+
+
+# Projects can define whatever they need to inside their config objects. In this
+# case, we're storing a `litellm_model_name`, which we can use to point
+# inference to a specific model on litellm, as well as a `use_thinking` flag,
+# which lets us compare models trained (or prompted) to use CoT to models
+# trained to just output the answer directly in the same project. Our rollout
+# and reward functions will use this config to adjust their behavior, but to ART
+# itself it's completely opaque.
+class MyConfig(BaseModel):
+    use_thinking: bool = False
+    # When using LightLLM / LiteLLM gateways you may want to override the
+    # underlying model name that is sent to the backend.
+    litellm_model_name: str | None = None
+
+
+# Dummy reward function that simply checks whether the model's answer is
+# correct. It supports both thinking and non-thinking models.
+def reward(config: MyConfig, scenario: MyScenarioSpec, response: str):
+    if config.use_thinking:
+        answer = re.search(r"<answer>(.*?)</answer>", response, re.DOTALL)
+        if answer is None:
+            return -1
+        response = answer.group(1).strip()
+    parsed_response = int(response)
+    if parsed_response == scenario.output:
+        return 1
+    else:
+        return -1
+
+
+# This function runs our rollout against our test scenarios, and logs the results.
+# We can use it to benchmark both trainable and prompted models.
+async def benchmark_model(model: art.Model):
+    trajectories = await art.gather_trajectories(
+        (rollout(model, scenario) for scenario in test_scenarios),
+        pbar_desc="benchmark",
+        max_exceptions=100,
+    )
+    valid_trajectories = [t for t in trajectories if isinstance(t, art.Trajectory)]
+    await model.log(valid_trajectories)
```

Comment:
I want to change `log` to also accept exceptions so that this kind of filtering isn't necessary. Also, being able to log exceptions will be valuable.

Reply:
yes agreed!